### ISSUE TITLE
Add automation to update the subset definitions in metadata.pb

### ISSUE THREAD

**Issue by MariannaPaszkowska** [googlefonts/googlesans-flex#1092]
We need to update our subset definitions, and ensure that we use a set of them that include every encoding present in the font—otherwise, missing encodings will be subset out. This is automated on the public families. This would help keeping the subset definitions in sync between the internal METADATA.pb (CSS API pipeline files) files and the ones in the repository. (thanks for advice here, @chrissimpkins )

@simoncozens mentioned that it requires a slight tweak to gftools-add-font, and then simply running it in the GSF directory. That should produce a METADATA.pb file with all the subsets.
Here is a PR prepared by Simon: https://github.com/googlefonts/gftools/pull/1065 that allows run gftools-add-font on a repo and get a GF-like METADATA.pb without being in a known libre license subdirectory.

- [ ]  @m4rc1e Could you have a look at how we can integrate this automated approach of defining that metadata file here?  

- [x] We also should consider promoting the above Font Bakery check from a WARN to a FAIL? (thanks for suggestion @Hoolean )

- [ ] @chrissimpkins  Below are the needed subsets I identified, but we should update it once we have an automated workflow for that.

```
latin, 
latin-ext 
vietnamese 
menu
math 
symbols 
```

**Comment by MariannaPaszkowska** [comment]
locally created metadata.pb via `gftools add-font`
only suggests including those subsets:
```
subsets: "latin"
subsets: "latin-ext"
subsets: "menu"
subsets: "vietnamese"
```

**Comment by chrissimpkins** [comment]
Hmm, so it thinks that the new sets you identified are not necessary? Can you provide an encoding from each of the math and symbols sets that you identified as not available in the other subset definitions? I can test for tofu with those encodings.  Ty!

**Comment by simoncozens** [comment]
As mentioned in the chat just now, the algorithm we use in gftools-add-font to decide whether or not a subset should be declared in METADATA.pb is pretty crude: if 50% of the subset's codepoints are in the font, the subset gets declared. This means that some very large subsets, containing lots of minority or rarely used codepoints, may not get declared even though a font supports a *language* perfectly well.

A good example is Devanagari. More than 50% of `devanagari-unique_glyphs.nam` is made up of Vedic signs, marks, annotations and so on; these need to be part of this subset - if they were in a "devanagari_ext" subset, then fonts which do support these Vedic signs won't be able to do mark attachment, because the marks and the bases would be split into two sub-fonts by the subsetter. But you don't need these esoteric codepoints to write Hindi, by any means.

The upshot is that using the `gftools-add-font` logic, any Devanagari font would not have the `devanagari` subset declared unless it did have all of these marks and signs.

I've been pushing for the threshold to be 1 glyph: if there's a supported codepoint in the font, it must be reachable from the web API and so a subset must be declared which "lets it out". But I admit that may not be practical.

Alternatively, a language-based approach might be better here: use shaperglot to analyse supported languages, and if there are any, then make sure that the subsets are declared which allow those languages to be rendered.

**Comment by MariannaPaszkowska** [comment]
I am now manually reviewing and deciding what subsets to add. Indeed, I run it with default threshold.

**Comment by m4rc1e** [comment]
> I've been pushing for the threshold to be 1 glyph: if there's a supported codepoint in the font, it must be reachable from the web API and so a subset must be declared which "lets it out". But I admit that may not be practical.

If we go this route, we get an absolute ton of subsets added to gs flex 

```
subsets: "adlam"
subsets: "ahom"
subsets: "armenian"
subsets: "avestan"
subsets: "balinese"
subsets: "bassa-vah"
subsets: "batak"
subsets: "bengali"
subsets: "bhaiksuki"
subsets: "brahmi"
subsets: "buginese"
subsets: "buhid"
subsets: "canadian-aboriginal"
subsets: "caucasian-albanian"
subsets: "chakma"
subsets: "cham"
subsets: "cherokee"
subsets: "coptic"
subsets: "devanagari"
subsets: "dogra"
subsets: "duployan"
subsets: "elbasan"
subsets: "gothic"
subsets: "grantha"
subsets: "gujarati"
subsets: "gunjala-gondi"
subsets: "gurmukhi"
subsets: "hanifi-rohingya"
subsets: "hanunoo"
subsets: "hebrew"
subsets: "javanese"
subsets: "kaithi"
subsets: "kannada"
subsets: "kayah-li"
subsets: "kharoshthi"
subsets: "khojki"
subsets: "khudawadi"
subsets: "lao"
subsets: "latin"
subsets: "latin-ext"
subsets: "lepcha"
subsets: "limbu"
subsets: "lisu"
subsets: "mahajani"
subsets: "malayalam"
subsets: "mandaic"
subsets: "manichaean"
subsets: "marchen"
subsets: "masaram-gondi"
subsets: "math"
subsets: "meetei-mayek"
subsets: "menu"
subsets: "meroitic"
subsets: "modi"
subsets: "mongolian"
subsets: "new-tai-lue"
subsets: "newa"
subsets: "nko"
subsets: "nushu"
subsets: "old-permic"
subsets: "oriya"
subsets: "osage"
subsets: "pahawh-hmong"
subsets: "phags-pa"
subsets: "psalter-pahlavi"
subsets: "rejang"
subsets: "saurashtra"
subsets: "sharada"
subsets: "siddham"
subsets: "sinhala"
subsets: "sogdian"
subsets: "sora-sompeng"
subsets: "soyombo"
subsets: "sundanese"
subsets: "syloti-nagri"
subsets: "symbols"
subsets: "syriac"
subsets: "tagalog"
subsets: "tagbanwa"
subsets: "tai-le"
subsets: "tai-tham"
subsets: "tai-viet"
subsets: "takri"
subsets: "tamil"
subsets: "telugu"
subsets: "thaana"
subsets: "thai"
subsets: "tifinagh"
subsets: "tirhuta"
subsets: "vietnamese"
subsets: "wancho"
subsets: "warang-citi"
subsets: "zanabazar-square"
```

**Comment by MariannaPaszkowska** [comment]
I don't think we should include all those. And even then, this leaves us with few codepoints not covered by any subsets.
The downside of setting the threshold so low is that AFAIK the API will cut a font for each of the subsets (did I get that part correct @m4rc1e ?) and that also can be very misleading, because the languages/scripts from those subsets will be reported as supported while this is not the case...

**Comment by m4rc1e** [comment]
>  and that also can be very misleading, because the languages/scripts from those subsets will be reported as supported while this is not the case...

Correct. the languages drop down on Google Fonts will end up showing this family.

![Image](https://github.com/user-attachments/assets/ebe9f9fb-920d-4464-a04b-cd995574e792)

**Comment by simoncozens** [comment]
> If we go this route, we get an absolute ton of subsets added to gs flex

Of course! Any subset with 0x25CC DOTTED CIRCLE (which needs to be in each subset so marks can attach to it) then gets magically declared. What a mess.

> Correct. the languages drop down on Google Fonts will end up showing this family.

I'm not sure that this is true. According to Rod's [language support doc](https://gist.github.com/rsheeter/169d4c60b3ab31c4cbbfd986623da9af), language support is computed by looking at the exemplar codepoints for each language in the lang repo, not based on subsets.

**Comment by MariannaPaszkowska** [comment]
I think including those:
```
subsets: "latin"
subsets: "latin-ext"
subsets: "menu"
subsets: "math"
subsets: "menu"
subsets: "symbols"
subsets: "vietnamese"
```
takes us already a long way.

Then we could add those, which can be misleading but will cover those glyphs:

```
subsets: "canadian-aboriginal"
subsets: "cherokee"
subsets: "syriac"
subsets: "nushu"
subsets: "tifinagh"
```

```
          • U+02C7 CARON: try adding one of: canadian-aboriginal, tifinagh, yi                                                            
          • U+02D8 BREVE: try adding one of: canadian-aboriginal, yi                                                                      
          • U+02D9 DOT ABOVE: try adding one of: canadian-aboriginal, yi                                                                  
          • U+02DB OGONEK: try adding one of: canadian-aboriginal, yi                                                                     
          • U+0306 COMBINING BREVE: try adding one of: old-permic, tifinagh                                                               
          • U+030A COMBINING RING ABOVE: try adding syriac                                                                                
          • U+030B COMBINING DOUBLE ACUTE ACCENT: try adding one of: cherokee, osage                                                      
          • U+030C COMBINING CARON: try adding one of: cherokee, tai-le  
          • U+2003 EM SPACE: try adding nushu    
```

and then we need to think what to do with glyphs that are not in any subsets. Marc also pointed out to me that some features are always subset out so just because glyphs are referenced in a feature does not quarantee that they will be left in :
```
    WARN The following codepoints supported by the font are not covered by any subsets defined in the font's metadata file, and will never be served. You can solve this by either manually adding additional subset declarations to METADATA.pb, or by editing the glyphset definitions.                
                                                                                           
          • U+02DD DOUBLE ACUTE ACCENT: not included in any glyphset definition                                                           
          • U+0312 COMBINING TURNED COMMA ABOVE: not included in any glyphset definition                                                  
          • U+031B COMBINING HORN: not included in any glyphset definition                                                                
          • U+0326 COMBINING COMMA BELOW: not included in any glyphset definition                                                         
          • U+0327 COMBINING CEDILLA: not included in any glyphset definition                                                             
          • U+0328 COMBINING OGONEK: not included in any glyphset definition                                                              
          • U+2004 THREE-PER-EM SPACE: not included in any glyphset definition                                                            
          • U+2005 FOUR-PER-EM SPACE: not included in any glyphset definition                                                             
          • U+2006 SIX-PER-EM SPACE: not included in any glyphset definition                                                              
          • U+200A HAIR SPACE: not included in any glyphset definition                                                                    
          • U+2070 SUPERSCRIPT ZERO: not included in any glyphset definition                                                              
          • U+2075 SUPERSCRIPT FIVE: not included in any glyphset definition                                                              
          • U+2076 SUPERSCRIPT SIX: not included in any glyphset definition                                                               
          • U+2077 SUPERSCRIPT SEVEN: not included in any glyphset definition                                                             
          • U+2078 SUPERSCRIPT EIGHT: not included in any glyphset definition                                                             
          • U+2079 SUPERSCRIPT NINE: not included in any glyphset definition                                                              
          • U+2080 SUBSCRIPT ZERO: not included in any glyphset definition                                                                
          • U+2081 SUBSCRIPT ONE: not included in any glyphset definition                                                                 
          • U+2082 SUBSCRIPT TWO: not included in any glyphset definition                                                                 
          • U+2083 SUBSCRIPT THREE: not included in any glyphset definition                                                               
          • U+2084 SUBSCRIPT FOUR: not included in any glyphset definition                                                                
          • U+2085 SUBSCRIPT FIVE: not included in any glyphset definition                                                                
          • U+2086 SUBSCRIPT SIX: not included in any glyphset definition                                                                 
          • U+2087 SUBSCRIPT SEVEN: not included in any glyphset definition                                                               
          • U+2088 SUBSCRIPT EIGHT: not included in any glyphset definition                                                               
          • U+2089 SUBSCRIPT NINE: not included in any glyphset definition                                                                
          • U+E000 : not included in any glyphset definition                                                                              
          • U+E001 : not included in any glyphset definition                                                                              
          • U+E002 : not included in any glyphset definition                                                                              
          • U+E003 : not included in any glyphset definition                                                                              
          • U+E004 : not included in any glyphset definition                                                                              
          • U+E005 : not included in any glyphset definition                                                                              
          • U+E006 : not included in any glyphset definition                                                                              
          • U+FB00 LATIN SMALL LIGATURE FF: not included in any glyphset definition                                                       
          • U+FB01 LATIN SMALL LIGATURE FI: not included in any glyphset definition                                                       
          • U+FB02 LATIN SMALL LIGATURE FL: not included in any glyphset definition                                                       
          • U+FB03 LATIN SMALL LIGATURE FFI: not included in any glyphset definition                                                      
          • U+FB04 LATIN SMALL LIGATURE FFL: not included in any glyphset definition   
                   
```

**Comment by MariannaPaszkowska** [comment]
@chrissimpkins Following our discussion, I added the subsets and the features here: https://github.com/googlefonts/googlesans-flex/pull/1095/commits/48283898f2331514369fe252ca7b77c3fc899807

**Comment by MariannaPaszkowska** [comment]
I also think we could improve the `unreachable_subsetting` fontbakery check:
```
com.google.fonts/check/metadata/unreachable_subsetting
Check for codepoints not covered by METADATA subsets.
```

Currently the check lists all the copepoints separately.
Would be good if it also suggests the fix in a better way - suggest a set of subsets that should be included to not subset any of the characters (if possible for those glyphs)
Example:
on top of that:
![Image](https://github.com/user-attachments/assets/25b93ed6-ba8f-4ede-b1c8-775414bbe8ba)

it would suggest also this:
```
add subsets: latin, math, symbols, 
```
basically suggesting those subsets with the most covered points and if a codepoint is already included in a more popular subset then it won't suggest adding it the less popular. 
@felipesanches @simoncozens 

I will post it to the fontbakery repo too.

**Comment by MariannaPaszkowska** [comment]
and I think the same could apply to the "gftools add-font " subset generation in the metadata.pb. It could look for the most optimal set of subsets. Maybe not always but there could be a flag for covering all codepoints with such logic (including most popular per glyph and overall a unique set)...

Like @m4rc1e said:
_(when) we want to cover them all but we want to do it by specifying the least subsets ;-)_

**Comment by chrissimpkins** [comment]
These are the subset definitions that we will use in the GSF v2.6 release that is already moving through the pipeline.  I plan to ship this to production by early December (likely the week following Thanksgiving week here in the US)

https://github.com/googlefonts/googlesans-flex/issues/1096

Please update that thread if you identify any further GSF v2.6 subset definition requirements as you work through the METADATA file prep automation in this issue. We'll use the automated METADATA file gen that Marianna is setting up for all future GSF releases after GSF v2.6.  Ty!

**Comment by MariannaPaszkowska** [comment]
A small tool by @m4rc1e to figure out the most optimal subsets for a given font: https://gist.github.com/m4rc1e/8fe31c10352a23cb8974cf25d1bcdca0
The next steps would be to add that to the build workflow and add tests based on this.
